paasta_tools/setup_namespace_services.py

```diff
@@ -0,0 +1,579 @@
+                if details is None:
+                    log.warning(
+                        f"Namespace {ns} in {service}/smartstack.yaml is empty, skipping"
+                    )
+                    continue
+                if not isinstance(details, Mapping):
+                    log.warning(
+                        f"Namespace {ns} in {service}/smartstack.yaml is not a mapping, skipping"
+                    )
+                    continue
+                namespaces[f"{service}.{ns}"] = details
+    except Exception as err:
+        log.warning(f"Failed to load namespaces for {service}: {err}")
+
+    return namespaces
+
+
+def load_smartstack_namespaces(
+    soa_dir: str = DEFAULT_SOA_DIR,
+    services: Optional[Set[str]] = None,
+) -> Mapping[str, Mapping]:
+    """Load smartstack namespace configurations from soa_dir.
+
+    Args:
+        soa_dir: Directory containing service configurations
+        services: Optional set of service names to load. If None, loads all services.
+
+    Returns:
+        Mapping of "service.namespace" to configuration dict
+    """
+    namespaces: Dict[str, Mapping] = {}
+
+    if services is not None:
+        # Load only specified services
+        for service in services:
+            namespaces.update(load_smartstack_namespaces_for_service(service, soa_dir))
+    else:
+        # Load all services
+        _, dirs, _ = next(os.walk(soa_dir))
+        for service_dir in dirs:
+            namespaces.update(
+                load_smartstack_namespaces_for_service(service_dir, soa_dir)
+            )
+
+    return namespaces
+
+
+def sanitise_kubernetes_service_name(name: str) -> str:
+    return limit_size_with_hash(sanitise_kubernetes_name(name).replace(".", "---"))
+
+
+def get_existing_kubernetes_service_names(
+    kube_client: KubeClient, kube_namespace: str
+) -> Set[str]:
+    try:
+        service_objects = kube_client.core.list_namespaced_service(
+            kube_namespace,
+            label_selector=f"{MANAGED_BY_LABEL}={FIELD_MANAGER}",
+        )
+    except ApiException as e:
+        if e.status == 404:
+            log.debug(
+                f"Namespace {kube_namespace} does not exist yet; assuming no managed Services"
+            )
+            return set()
+        raise
+
+    return {
+        item.metadata.name
+        for item in service_objects.items
+        if item.metadata and item.metadata.name
+    }
+
+
+def get_service_kube_namespace(service: str) -> str:
+    return f"paastasvc-{sanitise_kubernetes_name(service)}"
+
+
+def is_external_routing_enabled(config: Mapping) -> bool:
+    """Check if external routing is enabled for a namespace config.
+
+    Args:
+        config: Namespace configuration dict
+
+    Returns:
+        True if routing.external is set to True
+    """
+    routing_cfg = config.get("routing", {})
+    if not isinstance(routing_cfg, Mapping):
+        return False
+    return bool(routing_cfg.get("external"))
```

Comment:
imo, this might be better placed in the instance class (and we can probably assume that these are in the right format due to the schema, right?)